core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -1018,7 +1025,9 @@ private[spark] class BlockManager(
           try {
             replicate(blockId, bytesToReplicate, level, remoteClassTag)
           } finally {
-            bytesToReplicate.dispose()
+            if (!level.useOffHeap) {
```

Comment:
Allocating a direct byte buffer creates a `java.nio.DirectByteBuffer`, which is in turn a subclass of `java.nio.MappedByteBuffer`. So calling `dispose()` will dispose direct buffers, too.